### ISSUE TITLE
Notification App in Background or Closed

### ISSUE THREAD

**Issue by verissimomanoel** [fechanique/cordova-plugin-fcm#106]
When app is closed or running in background, the push callback doesn't get the data of push.

FCMPlugin.onNotification(
        function(data){
            if (data.wasTapped) {
                //Notification was received on device tray and tapped by the user.
                console.log(JSON.stringify(data)); **//Here not show anything** 

This problem only happen in android.

**Comment by ironbone** [comment]
I have almost the same problem.
The main problem is when I press the BACK Button and go out from the Application. The application is finished but still in the memory. The plugin does not recognize this situation. I can see in the console that the plugin receive the message but it does not display the notification and the application is not informed about this notification.
Additionality there is a similar problem with a background application. The plugin start first work well after some time. So when we send the application to the background we need to wait 10 - 20 sec. and then it works.

**Comment by aurbina83** [comment]
I have the same issue!  When not in the background, I do get "wasTapped: false" but when in background and tap on the notification bar, I get nothing. 

```
                if(data.wasTapped) {
                    alert("tapped");
                    alert( JSON.stringify(data));
                } 
```

**Comment by aurbina83** [comment]
I think there is a delay between setting the app back into foreground and connecting to FCM

**Comment by sur** [comment]
Had exactly the same issue in Android. Searched every documentation/discussion available but couldn't fix it.

Realised that the fix was rather simple and found it just by chance.

The solution is to:
just upgrade your Gradle to 2.14.1 and Android Studio to 2.2 (also upgrade gradle plugin to 2.2) and then build it.

**Comment by DetoloX** [comment]
I upgraded, and It didn't work for me..  :(

**Comment by aurbina83** [comment]
I've decided to just use OneSignal. Works flawlessly. Has "tags" which is similar to pubsub.

**Comment by sur** [comment]
@DetoloX I am sure you must have done, but just to confirm if after updating you have done "cordova platform update android" ?
or probably just removing the android platform and adding again ?

also closely monitor the Console Logs:
        FCMPlugin.onNotification(
            function(data){
              console.log('onNotification DATA received');

specifically check if you are having some logs after the function(data) ?

Sometimes, I also found the issue that even after fixing this bug (with gradle upgrade) I was getting the log for DATA correctly, but not getting the Push and that was due to incorrect settings for Whitelist plugin.

So please be sure about the situation with excessing logging.

And yeah, the last thing you need to make sure that you foreground push is working fine already ? right ? (I mean the token generation and setup on server side is all smooth).



@aurbina83 Yes, OneSignal is great and they have done a pretty good job and great thing is that for now just like Google they are also providing it for free for now.
But, the big advantage for me for using FCM is the great additional stuffs available in Firebase - like Crash Reporting and Analytics within one Google Firebase Console.
But if someone is purely focusing on implementing the Push on phonegap quickly, OneSignal is a good option today.

**Comment by aurbina83** [comment]
@sur I agree! OneSignal is a temporary solution. Once phonegap/fcm 2.0 is officially released I'm going to go back to FCM.

**Comment by DetoloX** [comment]
@sur Yes I updated correctly.. I received notifications when the app is in background  or when I am working with the app..

Mi problem is when the app is closed.. The device cann't received a notification..

In my server I send a notification like this, I try a hundred of differents combinations..

$fields = array(
			'to' => $registation_id,
			'priority' =>'high',
			'content_available' => true,
			'notification' => $notifiMSG,
			'data' => $dataMSG,
		);

I send notification and data values.. and It works great when the app is running..

**Comment by ironbone** [comment]
The same problem. I see at this moment the only solution - disable back button.

**Comment by esolutions4ev1** [comment]
In your notification payload make sure you have "click_action":"FCM_PLUGIN_ACTIVITY".  See if that works, it solved the problem for me.

**Comment by ironbone** [comment]
"click_action":"FCM_PLUGIN_ACTIVITY" does not solve the problem when the application is closed but still in the memory - for example after clicking the Back Button on your first page.

**Comment by esolutions4ev1** [comment]
@ironbone you are correct about that.  It does cover the other circumstances when in the background or closed which is the issue I was having.

**Comment by DetoloX** [comment]
@sur I try your solution in various android devices.. and the problem was on my Xiaomi Redmi Note 3, it has some kind of battery safe that kills the service or something like that.

I try in others devices and it works perfectly even when you reboot your device you can get notifications. :)

So people have a look on your device maybe has some kind of restriction your OS..

But in IOS what we can do??.. when the app is closed doesn't work.

**Comment by tobika** [comment]
Hello,
I have the same problem. Notifications work well in foreground. But once I open the app via the notification (in background or completely closed) I receive no notification callback at all

I updated to cordova-android@6.0 but this didn't change anything,
neither does FCM_PLUGIN_ACTIVITY

Do you think this is a plugin problem or specific android device problem?

Thanks for any hints and for the otherwise great plugin :)

**Comment by thecombo** [comment]
When I test the app from here https://cordova-plugin-fcm.appspot.com/  everything works fine.
But when I send notifications from my server PHP I have a problem when assigning or not assigning "notification" 

Case 1 : If the app is close and send 
 $fields = array('notification' => array("title" => "Test","body" => "Test Message","icon" => "default","sound" =>...... 
The notification appears on the bar but not when you open the app

Case 2: if I do not send this: 
 $fields = array('notification' => array("title" => "Test","body" => "Test Message","icon" => "default","sound" =>...... 
The notification not appears on the bar but yes when you open the app

**Comment by thecombo** [comment]
already resolve my problem. greetings developers

"priority":"high", //If not set, notification won't be delivered on completely closed iOS app
and 
"click_action":"FCM_PLUGIN_ACTIVITY",  //Must be present for Android

like this: 

//POST: https://fcm.googleapis.com/fcm/send
//HEADER: Content-Type: application/json
//HEADER: Authorization: key=AIzaSy*******************
{
  "notification":{
    "title":"Notification title",  //Any value
    "body":"Notification body",  //Any value
    "sound":"default", //If you want notification sound
    "click_action":"FCM_PLUGIN_ACTIVITY",  //Must be present for Android
    "icon":"fcm_push_icon"  //White icon Android resource
  },
  "data":{
    "param1":"value1",  //Any data to be retrieved in the notification callback
    "param2":"value2"
  },
    "to":"/topics/topicExample", //Topic or single device
    "priority":"high", //If not set, notification won't be delivered on completely closed iOS app
    "restricted_package_name":"" //Optional. Set for application filtering
}

**Comment by tobika** [comment]
Ok, I made a mistake with the "click_action":"FCM_PLUGIN_ACTIVITY" implementation on the server :-(
The notification in the background is well received now in the onNotification callback. :-)

But if the app is really closed (not in cache) I don't receive the callback. Does anybody have the same problem?

**Comment by kamilo81** [comment]
Any solution ?

When i receive background notification and not tap on it , notification doesnt appear in application. How to solve IT?

**Comment by MeteorStartup** [comment]
@tobika I got same issue. I can't get the callback when the app was actually closed by exit() function. 
it just process 'onNotification callback successfully registered' message (second argument of onNotification function) and do not trigger data callback.

is there any solution?

**Comment by tobika** [comment]
@MeteorStartup The event was sent after all (stupid me, didn't see it in the logs) but my app wasn't ready yet to take action yet (still loading). Maybe you have the same problem?

I just put the event in an global variable and checked when my app was ready if something was received.

**Comment by krishnanv001** [comment]
i am create cordova app using fcm plugin works fine.When user exit app while receive notification in device successfully.But Notification click cannot trigger onNotification() method to open specific page.It shows app home page using Android.

**Comment by romanarrieta** [comment]
I need a notification with click_action, but in cordova, where is name of activity?


I want to do the following:

......
notification":{
        "title":"Notification title",
        "body":"Notification body",
        "click_action":"hello.html",

    },
....

Is it possible??
Thank you
